### PR TITLE
simplify_exprt::simplify_bitwise is now independent of bitvector representation

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by tautschnig — COMMENTED** [review]
I think a unit test that evaluates the performance differences is necessary to show that it does not have a negative impact. The code path modified here might not be as hot as #3106, but care must be taken. Taking apart `mp_integer`s is very expensive. The fast path in `binary2integer` is in place for good reasons.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: eb0a4576a346538c5439aaedac6c66c4bd3ea1df).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87102947

**Comment by kroening** [comment]
Now with new approach that works on the representation directly!

File: src/util/arith_tools.cpp [hunk 1]
```diff
@@ -262,3 +262,39 @@ void mp_max(mp_integer &a, const mp_integer &b)
   if(b>a)
     a=b;
 }
+
+/// perform a binary bit-wise operation, given as a functor,
+/// on a bit-vector representation
+/// \param f: the functor
+irep_idt bitvector_bitwise_op(
```
Comment (tautschnig) [inline src/util/arith_tools.cpp:299]: Missing documentation of the other parameters, which trips up Doxygen.

File: src/util/arith_tools.h [hunk 2]
```diff
@@ -164,4 +164,19 @@ mp_integer power(const mp_integer &base, const mp_integer &exponent);
...
 
+/// perform a binary bit-wise operation, given as a functor,
+/// on a bit-vector representation
+irep_idt bitvector_bitwise_op(
+  const irep_idt &,
+  const irep_idt &,
+  const std::size_t width,
+  std::function<bool(bool, bool)>);
+
+/// perform a unnary bit-wise operation, given as a functor,
```
Comment (tautschnig) [inline src/util/arith_tools.h]: Typo: s/unnary/unary/, but also this documentation should go with the definition, not the declaration.

File: src/util/arith_tools.cpp [hunk 3]
```diff
@@ -262,3 +262,39 @@ void mp_max(mp_integer &a, const mp_integer &b)
...
+{
+  PRECONDITION(a.size() == width);
+  PRECONDITION(b.size() == width);
+
+  std::string result(width, ' ');
+
+  for(std::size_t i = 0; i < width; i++)
+  {
+    const bool bit_a = a[i] == '1';
+    const bool bit_b = b[i] == '1';
```
Comment (tautschnig) [inline src/util/arith_tools.cpp]: Should this become `get_bitvector_bit`?

File: src/util/arith_tools.cpp (hunk 1 above)
Comment (kroening) [reply src/util/arith_tools.cpp:299]: Fixed.

File: src/util/arith_tools.h (hunk 2 above)
Comment (kroening) [reply src/util/arith_tools.h]: fixed

File: src/util/arith_tools.cpp (hunk 3 above)
Comment (kroening) [reply src/util/arith_tools.cpp]: Yes, can do once #3106 is merged.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 855b35eb4f017dca4f113cb69f0c4ba4a2bf63d7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87130249

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 2a5599a685e1af6f7615f2ee144ba22894309892).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87134978
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 3db86441f08e7df9243f25e1f783ee156496968f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87139343

**Pull request merged**